blockpoller/poller.go

```diff
@@ -98,62 +119,178 @@ func (p *BlockPoller) run(resolvedStartBlock bstream.BlockRef) (err error) {
 	}
 
 	// On the first run, we will fetch the blk for the `startBlockRef`, since we have a `Ref` it stands
 	// to reason that we may already have the block. We could potentially optimize this
-	blk, err := p.fetchBlock(blkNum)
-	if err != nil {
-		return 0, fmt.Errorf("unable to fetch  block %d: %w", blkNum, err)
-	}
 
-	seenBlk, seenParent := p.forkDB.AddLink(blk.AsRef(), blk.ParentId, newBlock(blk))
+	seenBlk, seenParent := p.forkDB.AddLink(block.AsRef(), block.ParentId, newBlock(block))
 
-	currentState.addBlk(blk, seenBlk, seenParent)
+	currentState.addBlk(block, seenBlk, seenParent)
 
 	blkCompleteSegNum := currentState.getBlkSegmentNum()
-	blocks, reachLib := p.forkDB.CompleteSegment(blkCompleteSegNum)
+	completeSegment, reachLib := p.forkDB.CompleteSegment(blkCompleteSegNum)
 	p.logger.Debug("checked if block is complete segment",
 		zap.Uint64("blk_num", blkCompleteSegNum.Num()),
-		zap.Int("segment_len", len(blocks)),
+		zap.Int("segment_len", len(completeSegment)),
 		zap.Bool("reached_lib", reachLib),
 	)
 
 	if reachLib {
 		currentState.blkIsConnectedToLib()
-		err = p.fireCompleteSegment(blocks)
+		err := p.fireCompleteSegment(completeSegment)
 		if err != nil {
-			return 0, fmt.Errorf("firing complete segment: %w", err)
+			return 0, nil, fmt.Errorf("firing complete segment: %w", err)
 		}
 
 		// since the block is linkable to the current lib
 		// we can safely set the new lib to the current block's Lib
 		// the assumption here is that teh Lib the Block we received from the block fetcher ir ALWAYS CORRECT
-		p.logger.Debug("setting lib", zap.Stringer("blk", blk.AsRef()), zap.Uint64("lib_num", blk.LibNum))
-		p.forkDB.SetLIB(blk.AsRef(), blk.LibNum)
+		p.logger.Debug("setting lib", zap.Stringer("blk", block.AsRef()), zap.Uint64("lib_num", block.LibNum))
+		p.forkDB.SetLIB(block.AsRef(), block.LibNum)
 		p.forkDB.PurgeBeforeLIB(0)
 
-		err := p.saveState(blocks)
+		err = p.saveState(completeSegment)
 		if err != nil {
-			return 0, fmt.Errorf("saving state: %w", err)
+			return 0, nil, fmt.Errorf("saving state: %w", err)
 		}
 
-		return nextBlkInSeg(blocks), nil
+		nextBlockNum := nextBlkInSeg(completeSegment)
+		return nextBlockNum, nil, nil
 	}
 
 	currentState.blkIsNotConnectedToLib()
-	return prevBlkInSeg(blocks), nil
+
+	prevBlockNum, prevBlockHash := prevBlockInSegment(completeSegment)
+	return prevBlockNum, prevBlockHash, nil
 }
 
-func (p *BlockPoller) fetchBlock(blkNum uint64) (blk *pbbstream.Block, err error) {
-	var out *pbbstream.Block
-	err = derr.Retry(p.fetchBlockRetryCount, func(ctx context.Context) error {
-		out, err = p.blockFetcher.Fetch(ctx, blkNum)
+type BlockItem struct {
+	blockNumber uint64
+	block       *pbbstream.Block
+	skipped     bool
+}
+
+func (p *BlockPoller) loadNextBlocks(requestedBlock uint64, numberOfBlockToFetch int) error {
+	p.optimisticallyPolledBlocks = map[uint64]*BlockItem{}
+
+	nailer := dhammer.NewNailer(10, func(ctx context.Context, blockToFetch uint64) (*BlockItem, error) {
```

Comment:
A comment should be made here to explain that we use the nailer for its linearity of emitting the block in order received.